filament/backend/src/vulkan/utils/CappedArray.h

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (C) 2024 The Android Open Source Project
+ *
+* Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef TNT_FILAMENT_BACKEND_VULKAN_UTILS_CAPPEDARRAY_H
+#define TNT_FILAMENT_BACKEND_VULKAN_UTILS_CAPPEDARRAY_H
+
+// An Array that will be statically fixed in capacity, but the "size" (as in user added elements) is
+// variable. Note that this class is movable.
+
+#include <utils/Panic.h>
+
+#include <array>
+
+namespace filament::backend::fvkutils {
+
+template<typename T, uint16_t CAPACITY>
+class CappedArray {
+private:
+    using FixedSizeArray = std::array<T, CAPACITY>;
+
+public:
+    using const_iterator = typename FixedSizeArray::const_iterator;
+    using iterator = typename FixedSizeArray::iterator;
+
+    CappedArray() = default;
+
+    // Delete copy constructor/assignment.
+    CappedArray(CappedArray const& rhs) = delete;
+    CappedArray& operator=(CappedArray& rhs) = delete;
+
+    CappedArray(CappedArray&& rhs) noexcept {
+        this->swap(rhs);
+    }
+
+    CappedArray& operator=(CappedArray&& rhs) noexcept {
+        this->swap(rhs);
+        return *this;
+    }
+
+    inline ~CappedArray() {
+        clear();
+    }
+
+    inline const_iterator begin() const {
+        if (mInd == 0) {
+            return mArray.cend();
+        }
+        return mArray.cbegin();
+    }
+
+    inline const_iterator end() const {
+        if (mInd > 0 && mInd < CAPACITY) {
```

Comment:
`mInd` should never be equal to `CAPACITY`, so that should be an assert.

Reply:
Added an assert.